### ISSUE TITLE
Add a section for checking if the server responds with sensitive information.

### ISSUE THREAD

**Issue by pinkLagoon** [OWASP/wstg#727]
**What would you like added?**
We sometimes see that developers might just serialize an entire object into JSON and returns it in a API response. Sometimes this object contains sensitive information. I personally have seen passwords and API keys being displayed. I think this will fall under the new 4.12 API Testing section.

Would you like to be assigned to this issue?
I might later. If someone beats me to it that will be good.

**Comment by ThunderSon** [comment]
I highly believe this goes beyond API testing.
I just noticed the guide doesn't tackle that point, CWE 200 and similar vectors.
@kingthorin what do you think about this? Top 10 A3 (sensitive data exposure) is just referenced when data is sent unencrypted, but the guide in no way discusses "more" data being sent back to the user.
Would be interesting to find a home for it.

Nice catch @pinkLagoon!

**Comment by kingthorin** [comment]
Yup this is definitely something we need to tackle. I've definitely seen apps in the past that returned JSON or XML responses that contained more data than displayed to the user.

**Comment by ThunderSon** [comment]
This is currently being reviewed to see on how many stages this can be added.
1. Potentially dangerous data
2. Benign data -- cultural and leads to maturity

Talking with @elarlang ASVS doesn't have clear directions on that topic as well.

**Comment by elarlang** [comment]
> Talking with @elarlang ASVS doesn't have clear directions on that topic as well.

The solution in ASVS - added a requirement in the "defensive coding" section:
> Verify that the application returns only the data the user has permission to access. For example, the API response does not return a full object with attributes that contain values the user has no permission to access although having permission to access the data object.

**Comment by 0xBassia** [comment]
Hey, I'd like to take this one on. I do pentesting and bug bounty work and run into this constantly, APIs returning full serialized objects with fields the client never needs to see.

I'm thinking this fits as a new section under 4.12 API Testing, covering:
- APIs that serialize entire backend objects into responses without field filtering
- Hidden sensitive fields in JSON/XML responses (passwords, tokens, internal IDs, PII)
- Comparing what the UI displays vs what the API actually returns
- Error responses leaking internal data (stack traces, debug info, internal paths)
- GraphQL field-level over-exposure through permissive schemas

Will map it to CWE-200 and OWASP API Security Top 10 API3 (Broken Object Property Level Authorization).

I'll have a PR up soon.

**Comment by kingthorin** [comment]
Please don't reference CWE-200, it isn't meant to be mapped against.

![image](https://github.com/user-attachments/assets/19a4dd28-5a13-49c1-8e0e-6f361a3bc01b)
